source/articles/2016/use-an-underscore-when-memoizing-in-ruby.html.md

```diff
@@ -0,0 +1,128 @@
+---
+title: Use an Underscore When Memoizing in Ruby
+description: Let other developers know you don't want them to use your internals
+
+author: Adarsh Pandit
+category: writing
+date: 2016-10-12
+logo: ruby_on_rails_logo.png
+
+published: true
+tags: [ruby, rails, style, memoization, underscore, instance variable]
+---
+
+
+## TLDR;
+
+Do this:
+
+```ruby
+def expensive_method
+  @_expensive_method ||= #something
+end
+```
+
+Not this:
+
+```ruby
+def expensive_method
+  @expensive_method ||= #something
+end
+```
+
+![memoization](memo_pad.png)
+
+## Memoization Basics
+
+The act of "memoizing" a method means to
+locally cache the result of an expensive operation
+on the first method invocation
+and reusing it thereafter.
+
+It is functionally equivalent to:
+
+```ruby
+def expensive_method
+  @expensive_method = @expensive_method || this_expensive_operation
+end
+```
+
+## Naming
+
+By convention,
+you typically see developers name the instance variable after the method.
+Likely, this is because manual instance variable accessors
+used to look like this:
+
+```ruby
+def expensive_method
+  @expensive_method
+end
+```
+
+Nowadays, we usually write `attr_reader :expensive_method`.
+
+## Unused Variables
+
+Keen readers will note
+we are actually instantiating an instance variable here.
+That means we can access it from outside the class
+(assuming it's public):
+
+```ruby
+irb(main):001:0> def method_name; @ivar ||= 207; end
+:method_name
+irb(main):002:0> @ivar
+nil
+irb(main):003:0> method_name
+207
+irb(main):003:0> @ivar
+207
+```
+
+This is inadvertently adding to the class's public API.
```

Comment:
Is this assuming that there's also an `attr_reader :ivar`? If there's no `attr_reader`, the ivar is private:

```ruby
class Example
  def hello
    @greeting = "hello"
  end
end

example = Example.new
example.hello
# test.rb:9:in `<main>': undefined method `greeting' for #<Example:0x007fd77b22f010 @greeting="hello"> (NoMethodError)
puts example.greeting
```

It might be helpful to clarify this, especially since methods defined on irb are on a sort of global object that doesn't exist outside of irb.